services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,716 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TupleSections              #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeSynonymInstances       #-}
+{-# LANGUAGE ViewPatterns               #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- | This is 'MockGundeck', a re-implementation of the 'Gundeck' monad with mock effects.  This has
+-- been developed to test 'Gundeck.Push.pushAll' and 'Gundeck.Push.Websocket.bulkPush', but can be
+-- extended.
+--
+-- Besides 'MockGundeck' and its instances, there are two important types: 'MockEnv' and
+-- 'MockState'.  'MockEnv' contains the system input to the test run; 'MockState' is manipulated by
+-- the mock implementations and can be examined to make sure all the expected effects have been
+-- caused.
+--
+-- This module is structured as follows: first, 'MockEnv' and 'MockState' are defined with their
+-- 'Arbitrary' generators.  Then, 'MockGundeck' and its instances are defined; then the actual mock
+-- effects are implemented.
+--
+-- There is a cascade of mockings: 'pushAll' can be run on the same input as 'mockPushAll', and
+-- outputs and states can be tested to be equal.  'pushAll' calls 'bulkPush', but in the @instance
+-- MonadPushAll MockGundeck@, you can either call the real thing or 'mockBulkPush'.
+module MockGundeck where
+
+import Imports
+import Control.Lens
+import Control.Monad.Catch
+import Control.Monad.Except
+import Control.Monad.State
+import Data.Aeson
+import Data.Id
+import Data.List1
+import Data.Misc ((<$$>), Milliseconds(Ms))
+import Data.Range
+import Data.String.Conversions
+import Data.UUID ()
+import Gundeck.Aws.Arn as Aws
+import Gundeck.Options
+import Gundeck.Push
+import Gundeck.Push.Native as Native
+import Gundeck.Push.Websocket as Web
+import Gundeck.Types
+import Gundeck.Types.BulkPush
+import System.Logger.Class as Log hiding (trace)
+import System.Random
+import Test.QuickCheck as QC
+import Test.QuickCheck.Instances ()
+
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as Aeson
+import qualified Data.HashMap.Lazy as HashMap
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Data.Vector as Vector
+import qualified Network.URI as URI
+
+
+----------------------------------------------------------------------
+-- env
+
+type Payload = List1 Aeson.Object
+
+data MockEnv = MockEnv
```

Comment:
I wonder whether all the code that generates and consumes maps can be simplified by moving to a different representation:

```haskell
data ClientInfo = ClientInfo
    { _mrNativeAddress :: Maybe (Address "no-keys")
    , _mrWSReachable :: Bool
    , _mrNativeReachable :: Bool
    }

data MockEnv = MockEnv
    { _meRecipients :: Map (UserId, ClientId) ClientInfo
    }

clients :: UserId -> Lens' (Map (UserId, ClientId) a) (Map ClientId a)
clients = ...
```

This is less efficient but probably easier to work with and we won't have many clients in mock Gundeck so the inefficiency hopefully doesn't matter.

I haven't actually tried rewriting the code to use this representation, though.

Reply:
Oh, and also it makes `MockEnv` always valid (at least for now).

Reply:
- your data model means that a client can be native reachable if it doesn't have an address.  this could be fixed, though:

```haskell
data ClientInfo = ClientInfo
    { _ciNativeAddress   :: Maybe (Address "no-keys", Bool{- reachable -})
    , _ciWSReachable     :: Bool
    }
  deriving (Eq, Show)

data MockEnv = MockEnv
    { _meRecipients :: Map Recipient ClientInfo
    }
  deriving (Eq, Show)
```
- validity is still not guaranteed by it: Address can still contain a different `UserId` or `ClientId` from the one in the `meRecipients` map key.
- i'm not sure whether it would make things easier or more complicated.  `MockEnv` is read-only, so there is no problem with keeping the redundant information in sync.

but i'll give it a try.  i'm also curious how it'll go.